zcash_primitives/src/transaction/builder.rs

```diff
@@ -394,7 +435,8 @@ impl<R: RngCore + CryptoRng> Builder<R> {
         utxo: OutPoint,
         coin: TxOut,
     ) -> Result<(), Error> {
-        self.transparent_inputs.push(&mut self.mtx, sk, utxo, coin)
+        self.mtx.vin.push(TxIn::new(utxo));
+        self.transparent_inputs.push(sk, coin)
```

Comment:
By separating the `self.mtx.vin` mutation from `self.transparent_inputs.push()` and unconditionally mutating it first, the builder will end up in a corrupted state if an invalid input is added. Make the `self.mtx.vin` mutation conditional on the success of `self.transparent_inputs.push()`.